academy/3-ibc/5-token-transfer.md

```diff
@@ -238,12 +187,103 @@ func (k Keeper) SendTransfer(
+
+// FungibleTokenPacketData defines a struct for the packet payload
+// See FungibleTokenPacketData spec:
+// https://github.com/cosmos/ibc/tree/master/spec/app/ics-020-fungible-token-transfer#data-structures
+message FungibleTokenPacketData {
+    // the token denomination to be transferred
+    string denom = 1;
+    // the token amount to be transferred
+    string amount = 2;
+    // the sender address
+    string sender = 3;
+    // the recipient address on the destination chain
+    string receiver = 4;
+    // optional memo
+    string memo = 5;
+}
+```
+
+<HighlightBox type="note">
+
+Note that an optional _memo_ field was recently added to the packet definition. More details on the motivation, use cases and consequences will follow shortly.
+
+</HighlightBox>
+
+### Receiving a transfer packet
+
+A relayer will then pick up the `SendPacket` event and submit a `MsgRecvPacket` on the destination chain.
+
+This will then trigger an `OnRecvPacket` callback that will decode a packet and apply the transfer token application logic:
+
+```go
+// OnRecvPacket implements the IBCModule interface. A successful acknowledgement
+// is returned if the packet data is successfully decoded and the receive application
+// logic returns without error.
+func (im IBCModule) OnRecvPacket(
+    ctx sdk.Context,
+    packet channeltypes.Packet,
+    relayer sdk.AccAddress,
+) ibcexported.Acknowledgement {
+    ack := channeltypes.NewResultAcknowledgement([]byte{byte(1)})
+
+    var data types.FungibleTokenPacketData
+    if err := types.ModuleCdc.UnmarshalJSON(packet.GetData(), &data); err != nil {
+        ack = channeltypes.NewErrorAcknowledgement("cannot unmarshal ICS-20 transfer packet data")
+    }
+
+    // only attempt the application logic if the packet data
+    // was successfully decoded
+    if ack.Success() {
+        err := im.keeper.OnRecvPacket(ctx, packet, data)
+        if err != nil {
+            ack = types.NewErrorAcknowledgement(err)
+        }
+    }
+
+    ctx.EventManager().EmitEvent(
+        sdk.NewEvent(
+            types.EventTypePacket,
+            sdk.NewAttribute(sdk.AttributeKeyModule, types.ModuleName),
+            sdk.NewAttribute(types.AttributeKeyReceiver, data.Receiver),
+            sdk.NewAttribute(types.AttributeKeyDenom, data.Denom),
+            sdk.NewAttribute(types.AttributeKeyAmount, data.Amount),
+            sdk.NewAttribute(types.AttributeKeyAckSuccess, fmt.Sprintf("%t", ack.Success())),
+        ),
+    )
+
+    // NOTE: acknowledgment will be written synchronously during IBC handler execution.
+    return ack
+}
+```
+
+Note how we redirect to the module keeper's `OnRecvPacket` method, and are constructing the acknowledgment to be sent back.
+
+### Acknowledging or timing out packets
+
+The reader is invited to try to find and analyze the code corresponding to this (remember that the place to start is the packet callbacks, usually defined in a file like `module_ibc.go` or `ibc_module.go`).
```

Comment:
Hmm, are people using `module_ibc.go`? `ibc_module.go` should be recommend for consistency

Reply:
Ignite CLI uses `module_ibc.go`